### ISSUE TITLE
kernel blocked with cephfs on kernel 4.3+

### ISSUE THREAD

**Issue by guilhem** [coreos/bugs#1092]
We are using `cephfs` on production and after a migration to kernel 4.3 on `beta` channel, kernel hang.

Here is a ticket open on ceph bug tracker: http://tracker.ceph.com/issues/14360

**Comment by crawford** [comment]
Crap. Thanks for the heads up.

**Comment by crawford** [comment]
@guilhem do you have an easy repro case for this? I'm having trouble triggering the hung worker.

**Comment by mischief** [comment]
i've just tested this on CoreOS alpha 949.0.0. i'm unable to reproduce this issue. here's what i've run:

```
docker run --rm -ti --net=host -v /etc/ceph:/etc/ceph -e MON_IP=X.Y.Z.115 -e CEPH_NETWORK=X.Y.Z.0/24 ceph/demo
```
```
core@priv-lab-5 / $ sudo -s
bash-4.2# umount /mnt
bash-4.2# mount -t ceph -o name=admin,secret=SECRET X.Y.Z.115:6789:/ /mnt
bash-4.2# cd /mnt
bash-4.2# touch a b c
bash-4.2# ls -l
total 0
-rw-r--r-- 1 root root 0 Feb 11 01:20 a
-rw-r--r-- 1 root root 0 Feb 11 01:20 b
-rw-r--r-- 1 root root 0 Feb 11 01:20 c
bash-4.2# dd if=/dev/zero of=/mnt/zeroes bs=1M count=100
100+0 records in
100+0 records out
104857600 bytes (105 MB) copied, 0.0600963 s, 1.7 GB/s
bash-4.2# sync
bash-4.2# sestatus
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             mcs
Current mode:                   permissive
Mode from config file:          permissive
Policy MLS status:              enabled
Policy deny_unknown status:     denied
Max kernel policy version:      30
bash-4.2# setenforce 1
bash-4.2# ls
a  b  c  zeroes
bash-4.2# ls -l zeroes 
-rw-r--r-- 1 root root 104857600 Feb 11 01:21 zeroes
bash-4.2# exit
core@priv-lab-5 / $ sha512sum /mnt/zeroes 
2798503c2c7b718799324122137bf30a562aad1bc04bbf343daad225a5fd0d1fd5d269843a01ab00d4f8d8c5ab34f8956065f9831ef7459e9c487e895099e956  /mnt/zeroes
core@priv-lab-5 / $
```

i'll see what happens on beta.

@guilhem if you have any tips for reproducing this, it would be very helpful. don't really know much about cephfs so i'm not sure my testing will be an appropriate way to reproduce what you do with ceph.

**Comment by mischief** [comment]
fyi the above test is with btrfs

```
core@priv-lab-5 / $ findmnt /
TARGET SOURCE    FSTYPE OPTIONS
/      /dev/sda9 btrfs  rw,relatime,seclabel,ssd,space_cache,subvolid=5,subvol=/
```

**Comment by guilhem** [comment]
@mischief problem happen with: `w,relatime,name=admin,secret=<hidden>,fsc`. And may be solved with `context=system_u:object_r:tmp_t:s0` I'm testing it right now.

**Comment by mischief** [comment]
aha, so FSCACHE is involved. another moving part :-)

**Comment by mischief** [comment]
@guilhem can you describe your setup in more detail? how do you configure selinux? what is the workload when this problem occurs?

**Comment by guilhem** [comment]
@mischief I followed http://tracker.ceph.com/issues/14360
I added `context="system_u:object_r:tmp_t:s0"` to my mount options.

**Comment by mischief** [comment]
@guilhem the original bug report says this occurs while using SELINUXTYPE=mls. how are you configuring MLS on coreos?

**Comment by guilhem** [comment]
@mischief for the record: I didn't set anything other than `w,relatime,name=admin,secret=<hidden>,fsc` or set a selinuxtype when the problem occurs.

**Comment by mischief** [comment]
i've found a reproducer for this issue on CoreOS 899.7.0:

```
docker run -d --net=host -v /etc/ceph:/etc/ceph -e MON_IP=127.0.0.1 -e CEPH_NETWORK=127.0.0.1/8 ceph/demo
SECRET=$(docker run --rm -v /etc/ceph:/etc/ceph --entrypoint=/usr/bin/ceph-authtool ceph/demo -p /etc/ceph/ceph.client.admin.keyring)
sudo mount -t ceph -o rw,relatime,name=admin,secret=$SECRET,fsc 127.0.0.1:/ /mnt
sudo chown :core /mnt
sudo chmod g+w /mnt
git -C /mnt clone https://anongit.gentoo.org/git/repo/gentoo.git
```

**Comment by guilhem** [comment]
such a "good" news :)

**Comment by mischief** [comment]
@guilhem i commented on the upstream issue. apparently, they are aware of the problem and it seems they are aware of the solution, but it sounds non-trivial so i do not think this will be fixed soon.

**Comment by carmstrong** [comment]
Hey folks, I just saw that the stable channel now has a 4.3.6 kernel, but unfortunately this breaks clusters for us because of the Ceph bug.

**Comment by mischief** [comment]
@carmstrong can you use the workaround described in the upstream ceph bug?

**Comment by jcollie** [comment]
@mischief the workaround appears to work only temporarily

**Comment by carmstrong** [comment]
> the workaround appears to work only temporarily

Ugh :( how does it manifest, @jcollie?

**Comment by aemr3** [comment]
I got no issues with this workaround. I only added `context="system_u:object_r:tmp_t:s0"` to mount options. I am on CoreOS beta channel, kernel version is 4.4.1.

**Comment by jcollie** [comment]
@carmstrong there's a kernel traceback and then any access to the cephfs drive hangs. AFAIKS only a reboot fixes it. I'm not sure why CoreOS has this problem, as Fedora 23 using a 4.4 kernel doesn't but then maybe there's a patch in the Fedora kernel that has a better workaround.

**Comment by mischief** [comment]
it looks like torvalds/linux@315f24088048a51eed341c53be66ea477a3c7d16 will fix this.

**Comment by mischief** [comment]
ok, small update with some good news and some bad news.

with the 4.6 kernel which has some ceph patches in it, the good news is i can't get the kernel to hang. 

the bad news is, according to upstream the old userspace tools will not work now. so, i tried to use the 'jewel' release of ceph tools in a docker container, but i can't get the ceph volume to actually mount yet (hangs in mount command).

if i find out something more i'll leave another note here.

**Comment by mischief** [comment]
i think right now the trouble with the test case is some error in the setup script of ceph/demo container, so i'll have to try to fix the entrypoint.sh to work with jewel release.

**Comment by mischief** [comment]
ok, using the test in https://github.com/coreos/bugs/issues/1092#issuecomment-187995856 i can't reproduce the kernel hang with 4.6 kernel. if anyone continues to have problems when CoreOS alpha is released with the 4.6 kernel, ping this issue.

**Comment by carmstrong** [comment]
Thanks for the hard work on this, @mischief!
